zedUpload/datastore_http_test.go

```diff
@@ -274,22 +279,88 @@ func testHTTPDatastoreRepeat(t *testing.T) {
 	} else {
 		t.Log("Running HTTP repeat test suite.")
 
+		// make a random file
+		infile := httpDownloadDir + "input"
+		if _, err := os.Stat(infile); err == nil {
+			if err := os.Remove(infile); err != nil {
+				t.Fatalf("unable to remove existing file %s %v", infile, err)
+			}
+		}
+		f, err := os.Create(infile)
+		if err != nil {
+			t.Fatalf("unable to create file %s %v", infile, err)
+		}
+		defer os.RemoveAll(infile)
+		size := 1024 * 1024 * 100
+		bufSize := 1024 * 1024
+		randReader := io.LimitReader(rand.New(rand.NewSource(time.Now().UnixNano())), int64(size))
+		for {
+			buf := make([]byte, bufSize)
+			_, err := randReader.Read(buf)
+			if err != nil && err != io.EOF {
+				t.Fatalf("unable to read from random reader %v", err)
+			}
+			if _, err := f.Write(buf); err != nil {
+				t.Fatalf("unable to write to file %s %v", infile, err)
+			}
+			if err == io.EOF {
+				break
+			}
+		}
+		f.Close()
+
+		// create the test server
+		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			http.ServeFile(w, r, infile)
+		}))
+		defer ts.Close()
+		u, err := url.Parse(ts.URL)
+		if err != nil {
+			t.Fatalf("unable to parse url %s %v", ts.URL, err)
+		}
+		rport, err := strconv.Atoi(u.Port())
+		if err != nil {
+			t.Fatalf("unable to parse port %s %v", u.Port(), err)
+		}
+		lport := 9999
+
+		// start the proxy
 		go func() {
-			err := newUnstableProxyStart(9999, 80, "cloud-images.ubuntu.com", 1024*1024*100, 40*time.Second, 100)
+			err := newUnstableProxyStart(lport, rport, u.Hostname(), uint64(size/2), 40*time.Second, 100)
 			if err != nil {
 				t.Error(err)
 			}
 		}()
 
-		status, msg := operationHTTP(t, httpDownloadDir+"repeat2", "releases/focal/release/ubuntu-20.04.2-preinstalled-server-riscv64.img.xz", "http://127.0.0.1:9999", "", zedUpload.SyncOpDownload, true)
+		outfile := httpDownloadDir + "repeat2"
+		status, msg := operationHTTP(t, outfile, "path/does/not/matter/with/fixed/server", fmt.Sprintf("http://%s:%d", "127.0.0.1", lport), "", zedUpload.SyncOpDownload, true)
 		if status {
 			t.Errorf("%v", msg)
 		}
-		hashSum, err := sha256File(httpDownloadDir + "repeat2")
+		// get the number of bytes transferred
+		stat, err := os.Stat(outfile)
+		if err != nil {
+			t.Fatalf("unable to stat file %s %v", outfile, err)
+		}
+		txsize := stat.Size()
+
+		// get a hash of the part of the file that was transferred
+		f, err = os.Open(infile)
+		if err != nil {
+			t.Fatalf("unable to open file %s %v", infile, err)
+		}
+		defer f.Close()
+		hasher := sha256.New()
+		if _, err := io.CopyN(hasher, f, txsize); err != nil {
```

Comment:
Why are you trying to check hash only for the part of file? Seems it leads to the false-positive test result. In my runs in and out file size are differ, which is incorrect, we want to see the whole file downloaded.

Reply:
The whole file didn't download. It only downloaded the first 50MB. I thought that's what it was trying to do.

The whole thing was triggered because the download from Ubuntu gave a 404. I realized that we should just set up our own test server.

Clearly the test is not right then; what should it be doing? What is the actual expected behaviour?

Reply:
Expected behavior is to block all traffic using unstable proxy (in its [params](https://github.com/lf-edge/eve-libs/blob/1b116e736f77c6fd0d3c010977cc94f919063480/zedUpload/datastore_http_test.go#L278) you can see when to start, for how long and drop percent) for some amount of time and check that downloader handles retry and succeed download.
I am not sure that the test is broken, please see my [attempt](https://github.com/lf-edge/eve-libs/pull/4#issuecomment-1708710284) to bisect changes of downloader.

Reply:
Let's see if I get this. The purpose of the test is to check that lost packets do not mess it up, that the downloader should be able to figure out what got and what it did not, and then retry for what it needs. The unstable proxy then causes these lost packets, forcing the retry. If so, then the hash of the final downloaded file should match the hash of the original input file. 

If that's the case, something is wrong, either in the proxy or the client, as it stops after the number I put in the test. 

Did I understand this correctly?

Reply:
Yes, your understanding is correct.
Only the part with `it stops after the number I put in the test` is not fully clear to me. Idea is to use as large file as enough to start packets discarding in proxy (withiut discarding the test is noop). But yes, in and out files must be equal eventually.

Reply:
I'll update the PR to include a comment describing this understanding in the test. 

The original file we were downloading became a 404. So I generated a random file of 100MB, then set the proxy "begin crazy behavior" point at ½ of that. That should be a valid test, correct? And so if it is only getting half, then either the proxy or the client is broken?

Reply:
No, test is not valid with your changes now, as you check hash only for the part of file. Please check hash for the whole files or compare sizes of in and out files.
If we will see the size or the whole file's hashes missmatch, then yes, something is definitely broken

Reply:
@giggsoff I just updated the test. Please check the comment on the function describing the test, as well as the test itself, to see that it reflects what you described above. If the test fails now, we can figure out what is broken.

Reply:
Probably adding `for 40 seconds` or something like this after `configured to drop packets after half the file has been transferred` (in section 4) will improve understanding, as we do not want to drop packets forever, we want to continue download after a while.

Reply:
OK, look at the comment now.

This exercise is good, will help anyone coming to this in the future.

Reply:
Comment looks great now, thank you!

Reply:
Thanks for explaining it. Now to figure out what is broken. 😬

Reply:
I added comments and a bit of logical cleanup to the proxy. Looking at the last time I ran it, I see this:

```
Closed (191 bytes sent, 52427967 bytes received, 53478400 bytes discard)
```

That seems to imply that it sent half and discarded half. Shouldn't it then have later sent them again? 

From the proxy's perspective, it basically:

1. Sends half the bytes (0-startFromBytes)
2. Discard for a time (40 seconds)
3. Sends the rest

But since it _always_ calls `src.Read(buf)` - discard just says what to do with it _after_ the `Read()` - doesn't it mean that it _never_ will actually send the rest? Isn't the client basically hanging out there for 40 seconds, waiting, then eventually gets an `io.EOF`, without ever getting the rest?

By what mechanism is the client actually recovering?

Reply:
The logic of recovering hiding inside of zedUpload: we try several [times](https://github.com/lf-edge/eve-libs/blob/1b116e736f77c6fd0d3c010977cc94f919063480/zedUpload/httputil/http.go#L223) and use inactivityTimer there to handle retry if no progress in time.

Reply:
What is it actually doing? In the context of the original request, it is requesting the data. That gets about half the bytes, then discards the rest for 40 seconds (which is more than enough time to get the rest of the file), so it should be complete. What does the client do then? Does it treat it as an end of the download (i.e. failed, at least partially), and then start a _new_ request to the proxy?

Reply:
Also, interestingly, I am not convinced it is getting the `io.EOF` from the receiver, which it should send after it read all of the file (sending half and discarding half). I added some `fmt.Printf()` statements for every 1024 bytes, as well as when the `io.EOF` is received, I get the following:

```
Read 1024 bytes total 52427967 discarded 53475328, discard true
Read 1024 bytes total 52427967 discarded 53476352, discard true
Read 1024 bytes total 52427967 discarded 53477376, discard true
    datastore_http_test.go:83: Update progress for path/does/not/matter/with/fixed/server: 52363264/0
.... (waits up to 40 seconds) ....
Limit ended at:  2023-09-07 12:50:18.451865532 +0000 UTC m=+41.426247690
.... waits and waits and waits ...
send read EOF
    datastore_http_test.go:368: expected 104857600 bytes, got 52427776
Closed (191 bytes sent, 52427967 bytes received, 53478400 bytes discard)
--- FAIL: TestHTTPDatastore (301.31s)
    --- FAIL: TestHTTPDatastore/Repeat (301.31s)
FAIL
FAIL    github.com/lf-edge/eve-libs/zedUpload   301.328s
FAIL
```

Should it not get the `io.EOF` from the receive once the total file is downloaded?

Reply:
> What is it actually doing? In the context of the original request, it is requesting the data. That gets about half the bytes, then discards the rest for 40 seconds (which is more than enough time to get the rest of the file), so it should be complete. 

Cannot understand the idea here. Why it should be completed if we did not pass any data through our proxy?

> What does the client do then? Does it treat it as an end of the download (i.e. failed, at least partially), and then start a _new_ request to the proxy, but with the Range set?

Once upon a time (in the time of [commit](https://github.com/lf-edge/eve/pull/2210/commits/240687aa059ead702ddb108878d8fd222935aea3)) the idea was to rely on inactivityTimer which will move download to the next attempt starting from the current progress (copiedSize) regardless of the reason of hang. But seems now we just [return](https://github.com/lf-edge/eve-libs/blob/main/zedUpload/httputil/http.go#L338) without retry.

Reply:
I made some progress. Since we now serve the file with an `httptest.Server`, I changed the handler to provide the instability. This makes it simpler to control.

The test fails properly now. And we can see that it gets called only once, so the retry is not working. Now we can see why.

Does this make sense @giggsoff ?

Reply:
Sorry, I cannot understand why it happens in a few seconds. If so, something wrong with the novel logic. We define 40 seconds of sending nothing, why it ends so fast without waiting? And how client receive `server send is complete` if we do not pass anything (and EOF as well) through the proxy?

Reply:
Forget the proxy. What behaviour should the client see? If it requests 100 MB file, should it get return code 200 followed by 50MB and http close? Or should it get 200 followed by 50MB and then hang for 40 seconds? And if the latter, what should it get after the 40 seconds, the close?

Reply:
The latter one. Good question about what to send/receive after 40 seconds. I think nothing. We just remove limit and return possibility to pass packets, but should not do anything.

Reply:
It has to do something. If not, what does "40 seconds" mean? The only possibilities I see are:

- Keep it open indefinitely
- keep it open for 40 seconds and then close the connection

Reply:
Well, unfortunately, I have no good answer. We want to imitate real outage, so probably the first option without any signs will be more interesting

Reply:
@deitch @giggsoff let me know if there are other test changes as a result of this discussion.

Reply:
@eriknordmark there are. I just am not sure what behaviour to add. I managed to simplify the test, but it is not giving the same behaviour as before. We are trying to get to the right behaviour for this.

Reply:
@giggsoff I updated it so that it now has a 330 second timeout. Since the inactivityTimer is hard-coded to 5 mins, that should do the right thing, but hanging for > timeout.

Except that it doesn't. 

```
=== RUN   TestHTTPDatastore
=== RUN   TestHTTPDatastore/Repeat
    datastore_http_test.go:294: Running HTTP repeat test suite.
time="2023-09-10T08:41:56Z" level=info msg="ListenAndServe got request"
    datastore_test.go:181: Limit starts at: 2023-09-10 08:41:56.526040016 +0000 UTC m=+0.669634988
    datastore_test.go:195: Finished reading before end time, waiting until: 2023-09-10 08:47:26.526040016 +0000 UTC m=+330.669634988
    datastore_http_test.go:81: Update progress for path/does/not/matter/with/fixed/server: 51642368/0
    datastore_http_test.go:360: expected 105906176 bytes, got 52396544
2023/09/10 08:47:01 httptest.Server blocked in Close after 5 seconds, waiting for connections:
  *net.TCPConn 0xc0001220d8 127.0.0.1:33895 in state active
    datastore_test.go:199: Limit ends at: 2023-09-10 08:47:26.526183796 +0000 UTC m=+330.669778781
    datastore_test.go:218: Closed (105906688 bytes read, 52397056 bytes sent, 53509632 bytes discarded)
--- FAIL: TestHTTPDatastore (330.69s)
    --- FAIL: TestHTTPDatastore/Repeat (330.69s)
FAIL
FAIL    github.com/lf-edge/eve-libs/zedUpload   330.722s
FAIL
```

You can see that it hangs for the 330 seconds, with the server even complaining about the blocked `Close()` (which is what it should do).

However, the retry does not happen. 

Does this mean that the _test_ is incorrect? Or that the test is correct, but the _client_ behaviour is wrong? I think it is the client, but I really am not sure.

Reply:
I vote that the client is not correct, see https://github.com/lf-edge/eve-libs/pull/4#issuecomment-1708710284

Reply:
Hopefully that is a good starting point.

Somewhat relatedly, it is a pity that the inactivity timer is hard-coded to 5 minutes. It makes the tests take far longer than they should. If we could get that to 15 or 30 seconds, this would be much better.

But I couldn't figure out how to pass those as a parameter without changing the entire structure of the exposed API. Any halfway decent ideas @giggsoff or @eriknordmark ?

Reply:
I think we need @rouming on this one, as the commit is his.

@rouming I will save you the headache of reading the whole thread here. The repeat test fails:

```
=== RUN   TestHTTPDatastore
=== RUN   TestHTTPDatastore/Repeat
    datastore_http_test.go:294: Running HTTP repeat test suite.
time="2023-09-10T08:41:56Z" level=info msg="ListenAndServe got request"
    datastore_test.go:181: Limit starts at: 2023-09-10 08:41:56.526040016 +0000 UTC m=+0.669634988
    datastore_test.go:195: Finished reading before end time, waiting until: 2023-09-10 08:47:26.526040016 +0000 UTC m=+330.669634988
    datastore_http_test.go:81: Update progress for path/does/not/matter/with/fixed/server: 51642368/0
    datastore_http_test.go:360: expected 105906176 bytes, got 52396544
2023/09/10 08:47:01 httptest.Server blocked in Close after 5 seconds, waiting for connections:
  *net.TCPConn 0xc0001220d8 127.0.0.1:33895 in state active
    datastore_test.go:199: Limit ends at: 2023-09-10 08:47:26.526183796 +0000 UTC m=+330.669778781
    datastore_test.go:218: Closed (105906688 bytes read, 52397056 bytes sent, 53509632 bytes discarded)
--- FAIL: TestHTTPDatastore (330.69s)
    --- FAIL: TestHTTPDatastore/Repeat (330.69s)
FAIL
FAIL    github.com/lf-edge/eve-libs/zedUpload   330.722s
FAIL
```

We simplified the test so that it has a delay longer than the hard-coded timeout in the http client (5 mins). However, it does not do the retry. I cannot figure out how the retry is supposed to work, or if it ever gets called. I see the loop, but in the end, it looks like it just returns from the func, so how would the inactivityTimer ever matter?

This [commit](https://github.com/lf-edge/eve-libs/commit/faaf4a4acfd389135f8346d5f01d075dc83859a8#diff-438b9bb8c5586aa8dc22614121cdc879f824fb4e290c270f5ff946367d72703f) appears to be the one that Petr says does pass.

Reply:
First of all let me add @christoph-zededa, there was a bug we dived into together. 
The main purpose of the commit you are referencing to is to fail faster. There were a couple of problems discovered by customers, when the whole download stuck for tens of minutes, because of various and not obvious retry loops here and there.

The crucial part of this change is that once download starts (http header receives, but not full body) we give up download on a first error and retry on the next network interface (which happens in eve, not in this library). So if you have only 1 interface in tests, then probably retry never starts.

Reply:
![IMG_20230516_180719](https://github.com/lf-edge/eve-libs/assets/121871801/9f45d425-5473-4ad8-97f8-35c1ec66a163)

The retry should be done in `execCmdGet`

Reply:
Are you saying that it should not retry? And that this test should be removed? It sounds like that, as you're saying that the retry on other interface happens in whatever uses this library (eve) and not the library itself.

Out of curiosity, why do we have all of this inactivity custom stuff? Why do we not just rely on the http client timeout?

Reply:
I don't know for sure, best to set a breakpoint there to check it it gets called.

Reply:
I don't understand. Breakpoint? What for?

Reply:
To check if it enters the retry-loop.

Reply:
It doesn't. I can tell based on the timing and the logs, and I have run it with breakpoints. Based on scenarios:

* if the connection is closed, but not all of the data is sent, it thinks it was and ends
* if the connection is not closed, and all of the data is not sent, it reaches the hard-coded internal timeout (5 minutes) and returns an error, but does not retry

Forget for a moment what it _does_; what is the _expected_ behaviour?

Reply:
The expected behavior is to fail faster, at least the fix was targeting this issue. 

Avi, what is the internal timeout? 5 minutes is a lot, and as I said this is not a first patch which reduces the wait time on fail path. Do you know where the 5 minutes is set and where we exactly sleep waiting for timeout expiration?

Reply:
Is 5 min this inactivityTimer? I will check the code later, once reach the laptop l.

Reply:
@rouming thanks for taking a deeper look at it. Yes, the 5 minutes is the `inactivityTimeout` [here](https://github.com/lf-edge/eve-libs/pull/4/files#diff-4a3ca89f1c1037aa18b0258facf633f93de2016bfb09906794e7cc7c88f4fd62L32).

> previously instead of return an error we did retry

This PR didn't change that; `httputil/http.go` is barely changed for this PR, not that part at all. Which is why we were so stumped; we didn't change it here.

I 100% am happy to put the fix into this PR (which is why we asked for your help). I just am not sure what should be the right fix here.

You wrote:

```
     for attempt in 7 attempts {
            resp, err = http.Do()  <-- this one fails reasonably fast and we retry 7 times, so overall delay is ~4 minutes in total
            if err {
                   // Retry
                  continue
            }
           
           inactivityTimer.Reset(inactivityTimeout) <-- sets 5 min timeout
           until_everything_read {
			written, err = io.CopyN(resp) <-- now this can stall for 5 minutes 
                        if err {
***                        return err <-- this is the change in the patch! so we fail after 5 min of wait without retry
                        }
            }
     }
```

Correct. The retries apply only to creating the http connection. Once that happens and returns a successful 200, the 7 attempts are out of play. After that, it then tries to copy the body down. If it gets an error at that point - whether  a network error or the innerCtx timeout error (caused by the `inactivityTimer`), then it just returns the error.

Either way, that will not do the `Range` test, which, at least in theory, this is supposed to test.

With all of that in hand, what should we do here? I see two possibilities:

1. The behaviour is correct and the test is wrong. Once this client gets a valid 200, it should try to download and if it fails at some point (e.g. timeout), it should return an error. Whatever calls this client should deal with it. Conclusion: remove this test.
2. The test is correct and the behaviour is wrong. Even after this client gets a valid 200 and is downloading, if it gets an error during download it should try again, applying the retries to that as well. I am not sure the Range part works for this, but we will know if we change it. Conclusion: make the download error cause a retry rather than a `return err`.

What is the right thing to do in your opinion?

Reply:
I submitted a PR here: https://github.com/lf-edge/eve-libs/pull/8
It targets various things. Please take a look. The main part is it reverts the commit in question, which is not a proper way of doing things.

Reply:
No comment from me. If that one is correct, then we can get that in, and then I can rebase this on that.

Reply:
Yeah, let's do that. CI will not clear there due to various issues which are fixed here. Just comment there that you ran `go test` and it passed, Erik can override and merge in as administrator, and then I can rebase this on that one.